performance/load_test_performance_analysis.py

```diff
@@ -0,0 +1,365 @@
+
+
+def perform_test(
+    func: t_.Callable[..., t_.Any],
+    config: t_.Dict[str, t_.Any],
+    cli: Client,
+    **payload: t_.Any,
+) -> t_.Dict[str, t_.Union[int, float]]:
+    # Number of concurrent requests to simulate load
+    response_times: t_.List[float] = []
+    errors: t_.List[Exception] = []
+
+    start_time = time.time()
+    elapsed_time = 0.0
+
+    # User ramp-up
+    num_users = 0
+    while elapsed_time < config["total_run_time"]:
+        num_users += config["ramp_up_rate"]
+        num_users = min(num_users, config["num_peak_users"])
+
+        with ThreadPoolExecutor(max_workers=num_users) as executor:
+            # Submit tasks to the executor
+            futures = [
+                executor.submit(
+                    func,
+                    cli,
+                    response_times,
+                    errors,
+                    config["min_wait_time"],
+                    config["max_wait_time"],
+                    **payload,
+                )
+                for _ in range(num_users)
+            ]
+
+            # Wait for all tasks to complete
+            for future in futures:
+                future.result()
+
+        elapsed_time = time.time() - start_time
+
+    print()
+    # Calculate performance metrics
+    total_responses = len(response_times)
+    avg_response_time = (
+        sum(response_times) / total_responses if total_responses > 0 else 0
+    )
+    min_response_time = min(response_times) if total_responses > 0 else 0
+    max_response_time = max(response_times) if total_responses > 0 else 0
+    throughput = total_responses / elapsed_time if elapsed_time > 0 else 0
+    error_rate = (len(errors) / (len(response_times) + len(errors))) * 100
+
+    logger.info("Performance Metrics:")
+    logger.info(f"Run Time: {elapsed_time}")
+    logger.info(f"Number of Response: {total_responses}")
+    logger.info(f"Average Response Time: {avg_response_time} seconds")
+    logger.info(f"Minimum Response Time: {min_response_time} seconds")
+    logger.info(f"Maximum Response Time: {max_response_time} seconds")
+    logger.info(f"Throughput: {throughput} requests per second")
+    logger.info(f"Error Rate: {error_rate}%")
+
+    # Store performance metrics in XML file
+    return {
+        "run_time": elapsed_time,
+        "num_response": total_responses,
+        "avg_response_time": avg_response_time,
+        "min_response_time": min_response_time,
+        "max_response_time": max_response_time,
+        "throughput": throughput,
+        "error_rate": error_rate,
+    }
+
+
+def perform_get_cycles_load_test(
+    config: t_.Dict[str, t_.Any], cli: Client
+) -> t_.Dict[str, t_.Any]:
+    # Fetch machine information
+    machine_type = cli.get_machine_type_names()[0]
+    machines = cli.get_machine_names(source_type=machine_type)
+    columns = cli.get_machine_schema(machines[0])["display"][:].to_list()
+
+    # Define query parameters
+    query = {
```

Comment:
This comment caries for all the test but if we have a config file anyways why not add the ability to configure queries to test?

Reply:
Done.